sys/net/netstats/netstats_neighbor.c

```diff
@@ -0,0 +1,421 @@
+            *out = stats[i];
+            found = true;
+            break;
+        }
+    }
+
+    _unlock(dev);
+    return found;
+}
+
+/* find the oldest inactive entry to replace. Empty entries are infinity old */
+static netstats_nb_t *netstats_nb_get_or_create(netif_t *dev, const uint8_t *l2_addr, uint8_t len)
+{
+    netstats_nb_t *old_entry = NULL;
+    netstats_nb_t *stats = dev->neighbors.pstats;
+
+    timex_t cur;
+    xtimer_now_timex(&cur);
+
+    for (int i = 0; i < NETSTATS_NB_SIZE; i++) {
+
+        /* Check if this is the matching entry */
+        if (l2_addr_equal(stats[i].l2_addr, stats[i].l2_addr_len, l2_addr, len)) {
+            return &stats[i];
+        }
+
+        /* Entry is oldest if it is empty */
+        if (stats[i].l2_addr_len == 0) {
+            old_entry = &stats[i];
+            continue;
+        }
+
+        /* Check if the entry is expired */
+        if ((netstats_nb_isfresh(&stats[i]))) {
+            continue;
+        }
+
+        /* Entry is oldest if it is expired */
+        if (old_entry == NULL) {
+            old_entry = &stats[i];
+            continue;
+        }
+
+        /* don't replace old entry if there are still empty ones */
+        if (old_entry->l2_addr_len == 0) {
+            continue;
+        }
+
+        /* Check if current entry is older than current oldest entry */
+        old_entry = netstats_nb_comp(old_entry, &stats[i], cur.seconds);
+    }
+
+    /* if there is no matching entry,
+     * create a new entry if we have an expired one */
+    if (old_entry) {
+        netstats_nb_create(old_entry, l2_addr, len);
+    }
+
+    return old_entry;
+}
+
+void netstats_nb_record(netif_t *dev, const uint8_t *l2_addr, uint8_t len)
+{
+    _lock(dev);
+
+    int idx = cib_put(&dev->neighbors.stats_idx);
+
+    if (idx < 0) {
+        DEBUG("%s: put buffer empty\n", __func__);
+        goto out;
+    }
+
+    DEBUG("put %d\n", idx);
+
+    if (len == 0) {
+        /* Fill queue with a NOP */
+        dev->neighbors.stats_queue[idx] = NULL;
+    } else {
+        dev->neighbors.stats_queue[idx] = netstats_nb_get_or_create(dev, l2_addr, len);
+        dev->neighbors.stats_queue_time_tx[idx] = xtimer_now_usec();
+    }
+
+out:
+    _unlock(dev);
+}
+
+/* Get the first available neighbor in the transmission queue
+ * and increment pointer. */
+static netstats_nb_t *netstats_nb_get_recorded(netif_t *dev, uint32_t *time_tx)
+{
+    netstats_nb_t *res;
+    int idx = cib_get(&dev->neighbors.stats_idx);
+
+    if (idx < 0) {
+        DEBUG("%s: can't get record\n", __func__);
+        return NULL;
+    }
+
+    DEBUG("get %d (%d left)\n", idx, cib_avail(&dev->neighbors.stats_idx));
+
+    res = dev->neighbors.stats_queue[idx];
+    dev->neighbors.stats_queue[idx] = NULL;
+
+    *time_tx = dev->neighbors.stats_queue_time_tx[idx];
+
+    return res;
+}
+
+__attribute__((unused))
+static uint32_t _ewma(bool fresh, uint32_t old_val, uint32_t new_val)
```

Comment:
If you provide a `static inline` version of this `netstats_nb_t` for old value, that then resolves protected the access to `etx`, `rssi`, etc, in the function here, you could call this function from within C-conditionals instead of CPP-conditionals `#if IS_USED(…)`.

Reply:
I mean it like this, in case my ramblings were unclear:

```C
static uint32_t _ewma(bool fresh, uint32_t old_val, uint32_t new_val)
{
    // ...
}

static inline void _ewma_etx(bool fresh, netstat_nb_t *nb, uint32_t new_etx)
{
#ifdef MODULE_NETSTATS_NEIGHBOR_ETX
    nb->etx = _ewma(fresh, nb->etx, new_etx);
#endif
}

static inline void _ewma_rssi(bool fresh, netstat_nb_t *nb, uint32_t new_rssi)
{
#ifdef MODULE_NETSTATS_NEIGHBOR_ETX
    nb->rssi = _ewma(fresh, nb->rssi, new_rssi);
#endif
}

// ...
```

Reply:
I don't really see the advantage - that's basically what `netstats_nb_update_etx()` is already (I could just as well move the two if conditions outside the `#ifdef` block)

Reply:
Lesser and smaller `#ifdef` or `#if IS_USED(…)` blocks is the advantage in my opinion.

Reply:
I think having the entire function `#ifdef`ed out if the module is not used makes it clearer that this is a no-op.

The block in `netstats_nb_update_rssi()` can't be made any smaller as `stats->rssi` is only available if the `module_netstats_neighbor_rssi` module is used.

Reply:
> I think having the entire function `#ifdef`ed out if the module is not used makes it clearer that this is a no-op.

Is that the goal? IMHO code that contains multiple lines `#ifdef`'d, potenially giving room for nested `#ifdef`'s in the future make the code unclear in general.

Reply:
From https://github.com/RIOT-OS/RIOT/pull/14448/files#r568863184

> > I think having the entire function `#ifdef`ed out if the module is not used makes it clearer that this is a no-op.
> 
> Is that the goal? IMHO code that contains multiple lines `#ifdef`'d, potenially giving room for nested `#ifdef`'s in the future make the code unclear in general.

Ping?

Reply:
If you insist - 4868fa38b92906e100e83ab1db274c57aee26ca0

Reply:
Ah, now I get what you meant 4441da78367556c0f4fce0fb346c4b9e1ce9cc3a